network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -114,367 +114,276 @@ func NewControlMsgValidationInspector(
-	c.tracker.StoreActiveClusterIds(clusterIDList)
-}
-
-// blockingPreprocessingRpc ensures the RPC control message count does not exceed the configured discard threshold.
-// Expected error returns during normal operations:
-//   - ErrDiscardThreshold: if control message count exceeds the configured discard threshold.
-//
-// blockingPreprocessingRpc generic pre-processing validation func that ensures the RPC control message count does not exceed the configured hard threshold.
-func (c *ControlMsgValidationInspector) blockingPreprocessingRpc(from peer.ID, validationConfig *p2pconf.CtrlMsgValidationConfig, controlMessage *pubsub_pb.ControlMessage) error {
-	if validationConfig.ControlMsg != p2pmsg.CtrlMsgGraft && validationConfig.ControlMsg != p2pmsg.CtrlMsgPrune {
-		return fmt.Errorf("unexpected control message type %s encountered during blocking pre-processing rpc, expected %s or %s", validationConfig.ControlMsg, p2pmsg.CtrlMsgGraft, p2pmsg.CtrlMsgPrune)
-	}
-	count := c.getCtrlMsgCount(validationConfig.ControlMsg, controlMessage)
+// inspectIWant inspects RPC iWant control messages. This func will sample the iWants and perform validation on each iWant in the sample.
+// Ensuring that the following are true:
+// - Each iWant corresponds to an iHave that was sent.
+// - Each topic in the iWant sample is a valid topic.
+// If the number of iWants that do not have a corresponding iHave exceed the configured threshold an error is returned.
+// Args:
+// - iWant: the list of iWant control messages.
+// Returns:
+// - DuplicateFoundErr: if there are any duplicate message ids found in any of the iWants.
+// - IWantCacheMissThresholdErr: if the rate of cache misses exceeds the configured allowed threshold.
+// - error: if any error occurs while sampling the iWants, all returned errors are benign and should not cause the node to crash.
+func (c *ControlMsgValidationInspector) inspectIWant(from peer.ID, iWants []*pubsub_pb.ControlIWant) error {
+	lastHighest := c.rpcTracker.LastHighestIHaveRPCSize()
 	lg := c.logger.With().
-		Uint64("ctrl_msg_count", count).
 		Str("peer_id", from.String()).
-		Str("ctrl_msg_type", string(validationConfig.ControlMsg)).Logger()
+		Uint("max_sample_size", c.config.IWantRPCInspectionConfig.MaxSampleSize).
+		Int64("last_highest_ihave_rpc_size", lastHighest).
+		Logger()
 
-	c.metrics.BlockingPreProcessingStarted(validationConfig.ControlMsg.String(), uint(count))
-	start := time.Now()
-	defer func() {
-		c.metrics.BlockingPreProcessingFinished(validationConfig.ControlMsg.String(), uint(count), time.Since(start))
-	}()
+	if len(iWants) == 0 {
+		return nil
+	}
+	sampleSize := uint(10 * lastHighest)
+	if sampleSize == 0 || sampleSize > c.config.IWantRPCInspectionConfig.MaxSampleSize {
+		// invalid or 0 sample size is suspicious
+		lg.Warn().Str(logging.KeySuspicious, "true").Msg("zero or invalid sample size, using default max sample size")
+		sampleSize = c.config.IWantRPCInspectionConfig.MaxSampleSize
+	}
```

Comment:
In light of my previous comment, this sampling logic can be cleaned from all `inspect.*` methods.